### PR TITLE
Magland/issue45 - simplification of compute resource

### PR DESCRIPTION
**Pull request by magland**

### CODE REVIEW

File: bin/hither-compute-resource [hunk 1]
```diff
@@ -1,97 +1,517 @@
...
+        job_handler=jh,
+        job_cache=jc
+    )
+    print(f'Starting compute resource: {compute_resource_id}')
+    cr.run()
+
+@click.command(help='Start an event stream server.')
+def start_event_stream_server():
+    config_fname = os.getcwd() + '/compute_resource.json'
+    config = ComputeResourceConfig(config_fname)
```
Comment (jsoules) [inline bin/hither-compute-resource:248]: Not `EventStreamServerConfig`?

File: bin/hither-compute-resource [hunk 2]
```diff
@@ -1,97 +1,517 @@
...
+    dirname = os.getcwd() + '/event-stream-server'
+    do_run_docker_container(
+        image='magland/eventstreamserver:0.1.0',
+        docker_run_opts=f"-p {host_event_stream_server_port}:8080 -v {dirname}:/event-stream-server"
+    )
+
+@click.command(help='Start a kachery server')
+def start_kachery_server():
+    config_fname = os.getcwd() + '/compute_resource.json'
+    config = ComputeResourceConfig(config_fname)
```
Comment (jsoules) [inline bin/hither-compute-resource:262]: Not `KacheryServerConfig`?

File: bin/hither-compute-resource [hunk 3]
```diff
@@ -1,97 +1,517 @@
...
+    )
+
+@click.command(help='Configure a new compute resource server or update an existing one.')
+def config():
+    config_fname = os.getcwd() + '/compute_resource.json'
+    if not os.path.isfile(config_fname):
+        if not inquirer.confirm(f'Create a new configuration file at {config_fname}?', default=True):
+            print('Aborting configuration.')
+            return
+        with open(config_fname, 'w') as f:
```
Comment (jsoules) [inline bin/hither-compute-resource:281]: OK to overwrite if it already exists?

File: bin/hither-compute-resource [hunk 4]
```diff
@@ -1,97 +1,517 @@
...
+
+    if host_kachery_server or host_event_stream_server:
+        print('You can now start the compute resource server by running the following in separate terminals from this working directory:')
+    else:
+        print('You can now start the compute resource server by running the following command:')
+    if host_kachery_server:
+        print('hither-compute-resource start-kachery-server')
+    if host_event_stream_server:
+        print('hither-compute-resource start-event-stream-server')
+    print('hither-compute-resource start')
```
Comment (jsoules) [inline bin/hither-compute-resource:416]: It'd be nice to have a way to get a reminder of these command lines, without having to go through configuration again. I find myself forgetting them a lot with labbox (fortunately I wrote them down in a text file)

File: setup.py [hunk 5]
```diff
@@ -9,11 +9,14 @@
...
     packages=setuptools.find_packages(),
+    include_package_data=True,
     scripts=[
         "bin/hither-compute-resource"
     ],
     install_requires=[
-        "pymongo"
+        "pymongo",
+        "click",
+        "inquirer"
```
[the remark this answers is not in the document]
Comment (magland) [reply setup.py:19]: yes, let's ditch windows support. I made this issue: https://github.com/flatironinstitute/hither/issues/58

File: hither/docker_inject_user/injectUser.Dockerfile [hunk 6]
```diff
@@ -0,0 +1,44 @@
+# Inject a user with the same uid/gid as the current host user
+#
+# Example Usage:
+#
+# The below script will create a new docker image called image-name-nonroot
+# that is the same as the image-name image, except with a new non-root
```
[the remark this answers is not in the document]
Comment (magland) [reply hither/docker_inject_user/injectUser.Dockerfile:6]: 1. Yes and no. Docker is actually not needed for the compute resource server. It's needed for the kachery server and the event stream server. And singularity cannot be substituted for those.

2. hmmm, I hadn't thought of that. There might be an issue there. But this injectUser thing is not used for hither jobs... just used to host kachery and eventstream servers. So I think it's not a problem in that regard. But yeah, if we do use this mechanism elsewhere we should be careful.

File: bin/hither-compute-resource [hunk 7]
```diff
@@ -1,97 +1,517 @@
 #!/usr/bin/env python
 
-from typing import Dict, Any
+from hither import computeresource
```
[the remark this answers is not in the document]
Comment (magland) [reply bin/hither-compute-resource:3]: oh, it just wasn't being used in that file. We should ultimately type everyting.

File: bin/hither-compute-resource [hunk 8]
```diff
@@ -1,97 +1,517 @@
...
+            raise Exception(f'Problem loading configuration file: {config_fname}')
+        self._config_fname = config_fname
+    def get(self, key, default=None):
+        if type(key) == list:
+            if len(key) == 0:
+                raise Exception('Invalid key', key)
+            elif len(key) == 1:
+                return self.get(key=key[0], default=default)
+            else:
+                a = self.get(key=key[:len(key) - 1], default=dict())
```
[the remark this answers is not in the document]
Comment (magland) [reply bin/hither-compute-resource:32]: Created new issue to change this throughout.

File: bin/hither-compute-resource (hunk 1 above)
Comment (magland) [reply bin/hither-compute-resource:248]: No, this is correct as is.

File: bin/hither-compute-resource (hunk 2 above)
Comment (magland) [reply bin/hither-compute-resource:262]: No, this is correct as is.

File: bin/hither-compute-resource (hunk 3 above)
Comment (magland) [reply bin/hither-compute-resource:281]: We know that it does not exist from `if not os.path.isfile(config_fname)`

File: bin/hither-compute-resource (hunk 4 above)
Comment (magland) [reply bin/hither-compute-resource:416]: I agree. Maybe it should be in the help text for when you just type `hither-compute-resource`?

Comment (magland) [reply bin/hither-compute-resource:416]: I created an issue for this.

**Pull request merged**